vision/spatial_transformer/spatial_transformer.jl

```diff
@@ -0,0 +1,193 @@
+# Spatial Transformer Network
+
+#=
+In this tutorial we'll build a spatial transformer network that will transform MNIST
+digits for classification by a convolutional network
+
+* [Spatial Transformer Networks](https://proceedings.neurips.cc/paper/2015/hash/33ceb07bf4eeb3da587e268d663aba1a-Abstract.html)
+
+=#
+
+using LinearAlgebra, Statistics
+using Flux, Zygote, CUDA
+using Flux: batch, onehotbatch, flatten, unsqueeze
+using MLDatasets
+using IterTools: partition
+using Plots
+using ProgressMeter
+using ProgressMeter: Progress
+## =====
+
+args = Dict(
+    :bsz => 64, # batch size
+    :img_size => (28, 28), # mnist image size
+    :n_epochs => 40, # no. epochs to train
+)
+
+## ==== GPU
+dev = has_cuda() ? gpu : cpu
+
+## ==== Data 
+train_digits, train_labels = MNIST(split=:train)[:]
+test_digits, test_labels = MNIST(split=:test)[:]
+train_labels = Float32.(Flux.onehotbatch(train_labels, 0:9))
+test_labels = Float32.(Flux.onehotbatch(test_labels, 0:9))
```

Comment:
```suggestion
train_labels = Flux.onehotbatch(train_labels, 0:9)
test_labels = Flux.onehotbatch(test_labels, 0:9)
```
These shouldn't require conversion because the cross entropy loss functions can understand one-hot arrays, but give it a try.

Reply:
I'm getting issues with scalar indexing when trying to move the SparseArrays to gpu?

Reply:
Can you post the full error + stacktrace? There shouldn't be any sparse arrays at play here.

Reply:
Oops, this runs fine if `spatial_transformer.jl` is run as a script, but errors on the REPL:

 train_loader = DataLoader((train_digits |> dev, train_labels |> dev), batchsize=args[:bsz], shuffle=true, partial=false)   
DataLoader{Tuple{CuArray{Float32, 3, CUDA.Mem.DeviceBuffer}, Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}}, Random._GLOBAL_RNG, Val{nothing}}(([0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0;;; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0;;; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0;;; … ;;; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0;;; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0;;; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0; … ; 0.0 0.0 … 0.0 0.0; 0.0 0.0 … 0.0 0.0], Bool[Error showing value of type DataLoader{Tuple{CuArray{Float32, 3, CUDA.Mem.DeviceBuffer}, Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}}, Random._GLOBAL_RNG, Val{nothing}}:
ERROR: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method. 
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.     
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:33
  [2] assertscalar(op::String)
    @ GPUArraysCore C:\Users\aresf\.julia\packages\GPUArraysCore\rSIl2\src\GPUArraysCore.jl:78
  [3] getindex
    @ C:\Users\aresf\.julia\packages\GPUArrays\EVTem\src\host\indexing.jl:9 [inlined]
  [4] getindex
    @ C:\Users\aresf\.julia\packages\Flux\KkC79\src\onehot.jl:46 [inlined]
  [5] isassigned(::Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}, ::Int64, ::Int64)
    @ Base .\abstractarray.jl:553
  [6] _show_nonempty(io::IOContext{Base.TTY}, X::AbstractMatrix, prefix::String, drop_brackets::Bool, axs::Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}})
    @ Base .\arrayshow.jl:438
  [7] _show_nonempty(io::IOContext{Base.TTY}, X::Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}, prefix::String)
    @ Base .\arrayshow.jl:410
  [8] show
    @ .\arrayshow.jl:486 [inlined]
  [9] show_delim_array(io::IOContext{Base.TTY}, itr::Tuple{CuArray{Float32, 3, CUDA.Mem.DeviceBuffer}, Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}}, op::Char, delim::Char, cl::Char, delim_one::Bool, i1::Int64, n::Int64)       
    @ Base .\show.jl:1244
 [10] show_delim_array
    @ .\show.jl:1229 [inlined]
 [11] show(io::IOContext{Base.TTY}, t::Tuple{CuArray{Float32, 3, CUDA.Mem.DeviceBuffer}, Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}})
    @ Base .\show.jl:1262
 [12] _show_default(io::IOContext{Base.TTY}, x::Any)
    @ Base .\show.jl:413
 [13] show_default
    @ .\show.jl:396 [inlined]
 [14] show
    @ .\show.jl:391 [inlined]
 [15] show(io::IOContext{Base.TTY}, #unused#::MIME{Symbol("text/plain")}, x::DataLoader{Tuple{CuArray{Float32, 3, CUDA.Mem.DeviceBuffer}, Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}}, Random._GLOBAL_RNG, Val{nothing}})        
    @ Base.Multimedia .\multimedia.jl:47
 [16] display(d::REPL.REPLDisplay{REPL.LineEditREPL}, mime::MIME{Symbol("text/plain")}, x::DataLoader{Tuple{CuArray{Float32, 3, CUDA.Mem.DeviceBuffer}, Flux.OneHotArray{UInt32, 10, 1, 2, CuArray{UInt32, 1, CUDA.Mem.DeviceBuffer}}}, Random._GLOBAL_RNG, Val{nothing}})
    @ OhMyREPL C:\Users\aresf\.julia\packages\OhMyREPL\oDZvT\src\output_prompt_overwrite.jl:8
 [17] display(d::REPL.REPLDisplay, x::Any)
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:271
 [18] display(x::Any)
    @ Base.Multimedia .\multimedia.jl:328
 [19] #invokelatest#2
    @ .\essentials.jl:716 [inlined]
 [20] invokelatest
    @ .\essentials.jl:714 [inlined]
 [21] print_response(errio::IO, response::Any, show_value::Bool, have_color::Bool, specialdisplay::Union{Nothing, AbstractDisplay})
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:293
 [22] (::REPL.var"#45#46"{REPL.LineEditREPL, Pair{Any, Bool}, Bool, Bool})(io::Any)
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:277
 [23] with_repl_linfo(f::Any, repl::REPL.LineEditREPL)
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:510
 [24] print_response(repl::REPL.AbstractREPL, response::Any, show_value::Bool, have_color::Bool)
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:275
 [25] (::REPL.var"#do_respond#66"{Bool, Bool, REPL.var"#77#87"{REPL.LineEditREPL, REPL.REPLHistoryProvider}, REPL.LineEditREPL, REPL.LineEdit.Prompt})(s::REPL.LineEdit.MIState, buf::Any, ok::Bool)
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:846
 [26] #invokelatest#2
    @ .\essentials.jl:716 [inlined]
 [27] invokelatest
    @ .\essentials.jl:714 [inlined]
 [28] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
    @ REPL.LineEdit C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\LineEdit.jl:2493
 [29] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
    @ REPL C:\Users\aresf\AppData\Local\Programs\Julia-1.7.3\share\julia\stdlib\v1.7\REPL\src\REPL.jl:1232
 [30] (::REPL.var"#49#54"{REPL.LineEditREPL, REPL.REPLBackendRef})()
    @ REPL .\task.jl:429

Reply:
Does it still happen if you disable OhMyREPL? I've not seen this before with the vanilla one.